pkg/controller/egress_router/egress_router_controller.go

```diff
@@ -0,0 +1,212 @@
+	}, nil
+}
+
+func (r EgressRouterReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	klog.Infof("Reconciling egressrouter.network.operator.openshift.io %s\n", request.NamespacedName)
+
+	obj := &netopv1.EgressRouter{}
+	err := r.mgr.GetClient().Get(context.TODO(), request.NamespacedName, obj)
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Info("Egress Router %s seems to have been deleted\n", request.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		klog.Error(err)
+		return reconcile.Result{}, err
+	}
+
+	// Check to see if we already know this object
+	existing := r.egressrouters[request.NamespacedName]
+	if existing != nil {
+		if !reflect.DeepEqual(obj.Spec, existing.spec) {
+			klog.Infof("Egress Router %s has changed, refreshing\n", request.NamespacedName)
+			delete(r.egressrouters, request.NamespacedName)
+			existing = nil
+		}
+	}
+
+	if existing == nil {
+		klog.Infof("Creating a new Egress Router")
+		// Set owner reference to the controller
+		boolTrue := bool(true)
+		EgressRouterOwnerReferences := []v1.OwnerReference{
+			{
+				APIVersion: "network.operator.openshift.io/v1",
+				Kind:       "EgressRouter",
+				Name:       obj.Name,
+				UID:        obj.UID,
+				Controller: &boolTrue,
+			},
+		}
+		err := r.ensureEgressRouter(manifestDir, request.Namespace, obj, EgressRouterOwnerReferences)
+
+		if err != nil {
+			klog.Error(err)
+			r.egressrouterErrs[request.NamespacedName] =
+				errors.Wrapf(err, "could not reconcile Egress Router %s", request.NamespacedName)
+			r.setStatus()
+			return reconcile.Result{}, err
+		}
+
+		r.egressrouters[request.NamespacedName] = existing
+	}
+
+	if err != nil {
+		klog.Error(err)
+		r.egressrouterErrs[request.NamespacedName] =
+			errors.Wrapf(err, "could not reconcile Egress Router %s", request.NamespacedName)
+		r.setStatus()
+		return reconcile.Result{}, err
+	}
+
+	klog.Infof("successful reconciliation")
+	delete(r.egressrouterErrs, request.NamespacedName)
+	r.setStatus()
+	return reconcile.Result{RequeueAfter: ResyncPeriod}, nil
+}
+
+// setStatus summarizes the status of all Egress Router objects and updates the statusmanager
+// as appropriate.
+func (r *EgressRouterReconciler) setStatus() {
+	if len(r.egressrouterErrs) == 0 {
+		r.status.SetNotDegraded(statusmanager.EgressRouterConfig)
+	} else {
+		msgs := []string{}
+		for _, e := range r.egressrouterErrs {
+			msgs = append(msgs, e.Error())
+		}
+		r.status.SetDegraded(statusmanager.EgressRouterConfig, "EgressRouterError", strings.Join(msgs, ", "))
+	}
+}
+
+func (r *EgressRouterReconciler) ensureEgressRouter(manifestDir string, namespace string, router *netopv1.EgressRouter, EgressRouterOwnerReferences []v1.OwnerReference) error {
+	out := []*uns.Unstructured{}
+	data := render.MakeRenderData()
+	data.Data["ReleaseVersion"] = os.Getenv("RELEASE_VERSION")
+	data.Data["EgressRouterNamespace"] = namespace
+	if isItValidCidr(string(router.Spec.Addresses[0].IP)) {
```

Comment:
one last question: Does the CRD ensure that this array can't be empty? Otherwise this will panic.

Reply:
it does, we also thought about that. Thanks for pointing it out.

Reply:
Well, it says that the array is required, but it can be empty. So a quick length check would be a good idea.

Then we're good to go.

Reply:
ya I did have a `is minLength=1 check` defined, but was told to remove it or something in one of the iterations since it was not standard.